tensorboard/data/provider.py

```diff
@@ -0,0 +1,223 @@
+    "ScalarDatum", ("step", "wall_time", "value")
+)
+class ScalarDatum(_ScalarDatum):
+  """A single datum in a scalar time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This
+      is a unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as
+      `float` seconds since epoch.
+    value: The scalar value for this datum; a `float`.
+  """
+  pass
+
+
+class RunTagFilter(object):
+  """Filters data by run and tag names."""
+
+  def __init__(self, runs=None, tags=None):
+    """Construct a `RunTagFilter`.
+
+    A time series passes this filter if both its run *and* its tag are
+    included in the corresponding whitelists.
+
+    Order and multiplicity are ignored; `runs` and `tags` are treated as
+    sets.
+
+    Args:
+      runs: Collection of run names, as strings, or `None` to admit all
+        runs.
+      tags: Collection of tag names, as strings, or `None` to admit all
+        tags.
+    """
+    self._runs = None if runs is None else frozenset(runs)
+    self._tags = None if tags is None else frozenset(tags)
+
+  @property
+  def runs(self):
+    return self._runs
+
+  @property
+  def tags(self):
+    return self._tags
+
+  def test(self, run, tag):
+    if self._runs is not None and run not in self._runs:
+      return False
+    if self._tags is not None and tag not in self._tags:
+      return False
+    return True
+
+
+class StepFilter(object):
+  """Filters data in a time series by step."""
+
+  def __init__(self, lower_bound, upper_bound):
+    """Construct a `StepFilter`.
+
+    Negative values for `lower_bound` or `upper_bound` indicate indices
+    from the end of the array, with the same semantics as Python slices.
+
+    It is valid for `upper_bound` to be smaller than `lower_bound`; in
+    this case, if `lower_bound` and `upper_bound` are of the same sign,
+    no data will pass this filter.
+
+    Args:
+      lower_bound: The minimum step value permitted by this filter,
+        inclusive. Integer.
+      upper_bound: The maximum step value permitted by this filter,
+        inclusive. Integer.
+    """
+    self._lower_bound = lower_bound
+    self._upper_bound = upper_bound
+
+  @property
+  def lower_bound(self):
+    return self._lower_bound
+
+  @property
+  def upper_bound(self):
+    return self._upper_bound
+
+  def resolve(self, max_step):
+    """Resolve an actual lower and upper bound for a given time series.
+
+    Args:
+      max_step: The highest step of any event in the time series.
+
+    Returns:
+      A tuple `(lower_bound, upper_bound)` of nonnegative step values.
+    """
+    return (
+        self._resolve_step(max_step, self._lower_bound),
+        self._resolve_step(max_step, self._upper_bound),
+    )
+
+  def _resolve_step(self, max_step, step):
+    if step >= 0:
+      return step
+    else:
+      return max(0, max_step - ~step)
```

Comment:
I had to look up `~` in python because it's so uncommonly used.  I get that it's elegant, but can we avoid making twos-complement a prerequisite to understanding this logic and just do `max_step + (step + 1)` or `max_step - (abs(step) - 1)`?  Or at least add a comment like `# invert step from -1 ... -∞ to 0 ... ∞`

Reply:
Fascinating. I thought that this _was_ the natural way to think about
slice indices: the sign bit determines the base (start/end), cleanly
separating the space into two equal-sized(!) parts with complementation
as the bijection. It’s not actually about the numeric values; it’s about
the bits. But apparently this perspective is not shared, so I’ll happily
add a comment.

Reply:
Hah, yes, I did concede the elegance of the bijective approach :)  Comment sounds good.